### ISSUE TITLE
Safari page occasionally hangs indefinitely 

### ISSUE THREAD

**Issue by chheller** [DevExpress/testcafe-hammerhead#1770]
When running tests against safari (both locally and on Browserstack), occasionally the browser will simply hang, with the loading bar partially loaded, for several dozen minutes to a totally indeterminate amount of time. Sometimes, given enough time, it will fix itself after no less than half an hour. Other times, it's run for hours and remained stuck. Attached is a screen shot of several dozen errors that pop up in the console almost immediately after the first page loads. Chrome 69, Firefox 62, Edge 17, and IE 11 (the latter two on Browserstack) don't exhibit this behavior at all. It only seems to occur after the first test has completed

It appears most of the errors are related to hammerhead and cookie-sync, and given #1486 as well as the fact that _occasionally_ the test resumes after 30 or more minutes, that it's somehow related to cookie-sync. Additionally, the test never disconnects due to a failed heartbeat, and continues to run, which causes our CI builds to timeout and fail

![screen shot 2018-09-27 at 5 12 33 pm](https://user-images.githubusercontent.com/2985885/46177499-a41c9200-c278-11e8-988a-7b745c1d7e80.png)

![screen shot 2018-09-27 at 5 05 26 pm](https://user-images.githubusercontent.com/2985885/46177456-746d8a00-c278-11e8-9810-38c6b6065036.png)

Safari 11.1.2 and Safari 12.0.0 and Safari 10.1.2 on Mac OSX 10.13.6

**Comment by miherlosev** [comment]
Hi @chheller 

Could you please run `TestCafe` with [--dev](http://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#--dev) option and provide us a screenshot of the DevTools console.

**Comment by chheller** [comment]
Running `testcafe safari tests/smoke.ts --dev` didn't really appear to change the logs that much, however I have noticed that whenever the browser hangs, the expanded errors are the only difference. All the errors about matching host, domain, and port occur even on successful runs.

![screen shot 2018-09-28 at 9 34 35 am](https://user-images.githubusercontent.com/2985885/46215727-b68fdd80-c303-11e8-98f7-cfe8aec23f8c.png)
![screen shot 2018-09-28 at 9 51 13 am](https://user-images.githubusercontent.com/2985885/46215896-1dad9200-c304-11e8-977b-ef3de0f44599.png)

**Comment by miherlosev** [comment]
TestCafe tracks JavaScript errors and unhanded Promise rejections on the tested page. If the unhanded Promise rejection occurs on the page without using TestCafe, you can turn off script error tracking using the [--skip-js-error](http://devexpress.github.io/testcafe/documentation/using-testcafe/command-line-interface.html#-e---skip-js-errors) option. It the error occurs only with TestCafe, it's a bug. To fix it, we need access to your application or a simple example in which this problem is reproduced.

**Comment by chheller** [comment]
Unfortunately I can't grant access to the application, which I understand makes debugging from your perspective kind of impossible. I finally found a work around (after believing that we would not be able to use Safari with testcafe at all) and may have incidentally found the root cause. 

My original test looked like 
```ts
test(t: TestController) {
    await t
      .typeText(email, user.email)
      .typeText(password, user.password)
      .click(submitButton)
      .expect(getCurrentUrl())
      .contains('/#/dashboard');
}
// getCurrentUrl
export const getCurrentUrl = ClientFunction(() => document.location.href);
```

However, I changed the assertion to wait for an element to appear and the bug seems to have gone away. 
```ts
test(t: TestController) {
  await t
        .typeText(email, user.email)
        .typeText(password, user.password)
        .click(submitButton)
        .expect(myAccount.visible)
        .ok();
}
```

So, as far as I can tell, some error was occurring during that client function that retrieves the url, and it appears that it's not being handled correctly. I was following an [example here](https://devexpress.github.io/testcafe/documentation/test-api/obtaining-data-from-the-client/examples-of-using-client-functions.html) to create a command to get the current url. 

I'll close the issue for now, because I don't have a way for you guys to replicate this particular error, however I wanted to follow up on the off chance someone else encounters an obscure issue like this.
